File: PAT_Basic_Level_by_perl/1081 检查密码/1081 检查密码.pl
use strict;
use warnings;

# =====================================================================
# 题目：1081 检查密码
# 难度：简单
# 知识点：字符串处理, 条件判断
# =====================================================================
#
# 【题目描述】
#   检查用户密码是否合法。合法密码须不少于 6 个字符，只能由英文字母、
#   数字和小数点 '.' 组成，且必须既有字母也有数字。
#   按不同情况输出对应的提示信息。
#
# 【实现原理】
#   1. 长度不足 6 输出"太短"。
#   2. 存在非法字符（非字母、非数字、非'.'）输出"太乱"。
#   3. 无字母输出"needs zi mu"；无数字输出"needs shu zi"；否则合法。
#
# 【输入样例】
#   5
#   123s
#   ...
# 【输出样例】
#   Your password is tai duan le.
#   ...
# =====================================================================

# 读取密码数量 N
my $n = <STDIN>; chomp $n;

for (1..$n) {
    my $s = <STDIN>;
    chomp $s;

    if (length($s) < 6) {
        # 密码太短
        print "Your password is tai duan le.\n";
    } elsif ($s =~ /[^A-Za-z0-9.]/) {
        # 存在非法字符（只允许字母、数字、小数点）
        print "Your password is tai luan le.\n";
    } elsif ($s !~ /[A-Za-z]/) {
        # 只有数字，缺少字母
        print "Your password needs zi mu.\n";
    } elsif ($s !~ /[0-9]/) {
        # 只有字母，缺少数字
        print "Your password needs shu zi.\n";
    } else {
        # 合法密码
        print "Your password is wan mei.\n";
    }
}

File: PAT_Basic_Level_by_go/1081 检查密码/1081 检查密码.go
// 题目：1081 检查密码
// 实现原理：读取 N 个密码字符串，依次检查其合法性。先判断长度是否 >= 6，
// 再逐字符扫描，检查是否含有数字（hasDigit）、字母（hasLetter）以及非法字符（illegal）。
// 根据检查结果输出对应的提示信息。

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// main 是程序入口，读取输入并逐条检查密码的合法性，输出判定结果。
func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	if !scanner.Scan() {
		return
	}
	N, _ := strconv.Atoi(strings.TrimSpace(scanner.Text())) // 读取密码条数
	for i := 0; i < N; i++ {
		if !scanner.Scan() {
			break
		}
		pwd := strings.TrimRight(scanner.Text(), "\r\n") // 去除行尾换行符
		// 检查长度是否达标
		if len(pwd) < 6 {
			fmt.Println("Your password is tai duan le.")
			continue
		}
		hasDigit, hasLetter, illegal := false, false, false
		// 逐字符检查密码内容
		for _, c := range pwd {
			if c >= '0' && c <= '9' {
				hasDigit = true
			} else if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') {
				hasLetter = true
			} else {
				illegal = true
			}
		}
		// 根据检查结果输出对应提示
		if !hasDigit {
			fmt.Println("Your password needs shu zi.")
		} else if !hasLetter {
			fmt.Println("Your password needs zi mu.")
		} else if illegal {
			fmt.Println("Your password is tai luan le.")
		} else {
			fmt.Println("Your password is wan mei.")
		}
	}
}
